### PR TITLE
Deprecate `github` formatter

### LINKED ISSUE #7447: Deprecate GitHub formatter
### What is the problem you're trying to solve?

Ref: https://github.com/stylelint/stylelint/issues/7430#issuecomment-1872329954

### What solution would you like to see?

Deprecate the GitHub formatter and remove it in the next major release.

Ideally, we'd link to a newly published GitHub custom formatter as part of the deprecation changelog entry.

### PR DESCRIPTION
**Pull request by marcalexiei**
> Which issue, if any, is this issue related to?

Close #7447

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory

### CODE REVIEW

File: .changeset/twenty-apes-shave.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+"stylelint": minor
+---
+
+Deprecate `github` formatter
+
+You can use [stylelint-actions-formatters](https://github.com/xt0rted/stylelint-actions-formatters) instead
```
Comment (Mouvedia) [inline .changeset/twenty-apes-shave.md:7]:
```suggestion
Deprecated: GitHub formatter
```

File: lib/standalone.mjs [hunk 2]
```diff
@@ -83,6 +83,16 @@ export default async function standalone({
 	let formatterFunction;
 
 	try {
+		/** @see https://github.com/stylelint/stylelint/issues/7447 */
+		if (isString(formatter) && formatter === 'github') {
+			process.emitWarning(
```
Comment (Mouvedia) [inline lib/standalone.mjs:88]: medium

Use `emitDeprecationWarning` util.

File: types/stylelint/index.d.ts [hunk 3]
```diff
@@ -188,6 +188,7 @@ declare namespace stylelint {
 	type Formatters = {
 		readonly compact: Promise<Formatter>;
 		readonly github: Promise<Formatter>;
+		/** @deprecated */
```
Comment (Mouvedia) [inline types/stylelint/index.d.ts:191]: It seems to be on the wrong line.

**Review by Mouvedia — CHANGES_REQUESTED** [review]
@marcalexiei 
Thanks for the PR.
I have made some comments, please review them.

File: lib/standalone.mjs (hunk 2 above)
Comment (marcalexiei) [reply lib/standalone.mjs]: Added a new code to `emitDeprecationWarning`:  `GITHUB_FORMATTER: '004'`

File: lib/standalone.mjs [hunk 4]
```diff
@@ -83,6 +84,15 @@ export default async function standalone({
 	let formatterFunction;
 
 	try {
+		/** @see https://github.com/stylelint/stylelint/issues/7447 */
+		if (isString(formatter) && formatter === 'github') {
```
Comment (Mouvedia) [inline lib/standalone.mjs:88]: question

Is `isString(formatter) &&` necessary?

File: lib/__tests__/cli-deprecation.test.mjs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { fileURLToPath } from 'node:url';
+import path from 'node:path';
+import process from 'node:process';
+
+import { expect, jest } from '@jest/globals';
```
Comment (Mouvedia) [inline lib/__tests__/cli-deprecation.test.mjs:5]: nit

remove `expect`

File: lib/standalone.mjs (hunk 4 above)
Comment (marcalexiei) [reply lib/standalone.mjs:88]: Actually no... even if it is a function or `undefined` the `isString` check is not necessary

**Comment by Mouvedia** [comment]
Unless you wanna add a skipped test to cover #7621, this looks good to me.

**Review by Mouvedia — APPROVED** [review]

**Comment by marcalexiei** [comment]
IMHO #7826 introduces a new config option so I think it's better to add the test there directly.

**Comment by marcalexiei** [comment]
Obviously If #7826 is merged before this PR I have no problem adding the tests here 😃

File: lib/standalone.mjs [hunk 6]
```diff
@@ -83,6 +84,15 @@ export default async function standalone({
...
 	try {
+		/** @see https://github.com/stylelint/stylelint/issues/7447 */
+		if (formatter === 'github') {
+			emitDeprecationWarning(
+				'"github" formatter is deprecated',
+				'GITHUB_FORMATTER',
+				'Please use "stylelint-actions-formatters" instead.',
+			);
+		}
+
```
Comment (Mouvedia) [inline lib/standalone.mjs:95]: minor
refactor

I think this should be moved in `getFormatterFunction` inside the first `if`.

File: lib/standalone.cjs [hunk 7]
```diff
@@ -303,6 +304,15 @@ async function standalone({
  */
 function getFormatterFunction(selected) {
 	if (typeof selected === 'string') {
+		/** @see https://github.com/stylelint/stylelint/issues/7447 */
+		if (selected === 'github') {
```
Comment (Mouvedia) [inline lib/standalone.cjs:308]: @marcalexiei
Thinking about it more, you will need to add `!quietDeprecationWarnings && ` to the condition, hence you should put that `if` where it was originally.
i.e. l88–94

---

@ybiquitous after this PR we should put the `if (!quietDeprecationWarnings)`inside the `emitDeprecationWarning` util.

**Review by Mouvedia — APPROVED** [review]

File: .changeset/twenty-apes-shave.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+---
+"stylelint": minor
+---
+
+Deprecated: GitHub formatter
```
Comment (ybiquitous) [inline .changeset/twenty-apes-shave.md:5]: [suggest] Super minor, but for consistency:

```suggestion
Deprecated: `github` formatter
```

**Review by ybiquitous — APPROVED** [review]
@marcalexiei Thanks for the first contribution. LGTM! 👍🏼

**Comment by ybiquitous** [comment]
Oh, I forgot the document: 😅 
https://github.com/stylelint/stylelint/blob/0850a28a1780c84ae55f179f60dcf0ea6971f4a6/docs/user-guide/options.md?plain=1#L87

@marcalexiei Can you also update it?

File: docs/user-guide/options.md [hunk 9]
```diff
@@ -84,7 +84,7 @@ Specify the formatter to format your results.
 Options are:
 
 - `compact` - generates output similar to ESLint's compact formatter
-- `github` - generates messages via [workflow commands for GitHub Actions](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions)
```
Comment (marcalexiei) [inline docs/user-guide/options.md:87]: @ybiquitous I added a `(DEPRECATED)` at the end of the string.
I also changed the same string in `cli.mjs` so it is more visible.

**Review by ybiquitous — APPROVED** [review]
Thanks. LGTM 👍🏼

**Pull request merged**